pcdscalc/be_lens_calcs.py

```diff
@@ -0,0 +1,368 @@
+import pprint
+from itertools import product
+
+
+"""
+If you are using the IMS IOC, each lens motor should have a
+base record associated with it  i.e MFX:DIA:XFLS.
+There should also be a binary record for each state,
+that is just this base plus an arbitrary string i.e MFX:DIA:XFLS:OUT
+"""
+
+# We have sets of Be lenses with thicknesses:
+LENS_RADII = [50e-6, 100e-6, 200e-6, 300e-6, 500e-6, 1000e-6, 1500e-6]
+
+
+def getAttLen(E, material="Be", density=None):
+    """ get the attenuation length (in meter) of material (default Si), if no
+      parameter is given for the predefined energy;
+      then T=exp(-thickness/att_len); E in keV"""
+    att_len = float(xsf.attenuation_length(material, density=density,
+                    energy=E))
+    return att_len
+
+
+def get_delta(E, material="Be", density=None):
+    """
+    Calculates delta for a given material at a given energy
+
+    Parameters
+    ----------
+    E: number
+        Beam Energy
+    material : `str`
+        Beryllium. The use of beryllium extends the range of operation
+        of compound refractive lenses, improving transmission,
+        aperture size, and gain
+    density : TODO: find out what density is
+    Returns
+    --------
+    delta: `int` TODO: not sure if int here or float?
+    """
+    delta = 1 - np.real(xsf.index_of_refraction(material, density=density,
+                        energy=E))
+    return delta
+
+
+def calc_focal_length_for_single_lens(E, radius, material="Be", density=None):
+    """
+    Calculates the Focal Length for a single lens.
+
+    Parameters
+    ----------
+    E: number
+        Beam Energy
+    radius : float TODO: is this float or no?
+    material : `str`
+        Beryllium. The use of beryllium extends the range of operation
+        of compound refractive lenses, improving transmission,
+        aperture size, and gain
+    density : TODO: find out what density is
+
+    Returns
+    --------
+    focal_length : `float`
+        The focal length for a single lens
+    """
+    delta = get_delta(E, material, density)
+    focal_length = radius / 2.0 / delta
+    return focal_length
+
+
+def calc_focal_length(E, lens_set, material="Be", density=None):
+    """
+    Calculates the Focal Length for certain lenses configuration and energy.
+
+    Parameters
+    -----------
+    E: number
+        Beam Energy
+    lens_set : `list`
+        [(numer1, lensthick1), (number2, lensthick2)...]
+    material : `str`
+        Beryllium. The use of beryllium extends the range of operation
+        of compound refractive lenses, improving transmission,
+        aperture size, and gain
+    density : TODO: find out what density is
+
+    Returns
+    -------
+    focal_length : `float`
+    """
+    num = []
+    rad = []
+    ftot_inverse = 0
+    # if type(lens_set) is int:
+    #     lens_set = getLensSet(lens_set)
+    for i in range(len(lens_set) / 2):
+        # TODO: why new code is (double //) range(len(lens_set)//2):
```

Comment:
`//` means floor division - https://docs.python.org/3/glossary.html#term-floor-division

It's appropriate here as Python 3 `range` will not take a `float`.

For fun, an alternative for this loop could be using numpy to reshape it:
```
In [7]: np.arange(20).reshape((10, 2))[:, 0]
Out[7]: array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18])

In [8]: np.arange(20).reshape((10, 2))[:, 1]
Out[8]: array([ 1,  3,  5,  7,  9, 11, 13, 15, 17, 19])

# then zip(_7, _8)
```

Considering it looks like there may be `None` in there, using Python to `zip` items together may be preferable:
```python
In [9]: items = np.arange(20)

In [10]: print(list(zip(items[::2], items[1::2])))
[(0, 1), (2, 3), (4, 5), (6, 7), (8, 9), (10, 11), (12, 13), (14, 15), (16, 17), (18, 19)]
```

Reply:
Great, thank you, I added the `//` back yesterday after reading about it (well after getting the error that we can't use range for flaot haha) - that was something I did not know we can do, but I like the the idea of using numpy - i'll give it a try. 

Also, there are a few places I noticed with possible having `None` - I will try this too - thanks a lot